drivers/gpio/gpio_davinci.c

```diff
@@ -0,0 +1,185 @@
+	struct gpio_driver_config common;
+	uint32_t port_base;
+	uint32_t port_num;
+};
+
+static int gpio_davinci_configure(const struct device *dev, gpio_pin_t pin,
+									gpio_flags_t flags)
+{
+
+	struct gpio_davinci_config *config = (struct gpio_davinci_config *)dev->config;
+
+	volatile struct gpio_davinci_regs *regs =
+				(struct gpio_davinci_regs *)config->port_base;
+
+	if ((flags & GPIO_SINGLE_ENDED) != 0) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) != 0) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & GPIO_OUTPUT) != 0) {
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			regs->set_data = (BIT(pin));
+		} else {
+			regs->clr_data = (BIT(pin));
+		}
+
+		regs->dir &= (~(BIT(pin)));
+
+	} else {
+		regs->dir |= (BIT(pin));
+	}
+
+	return 0;
+}
+
+static int gpio_davinci_port_get_raw(const struct device *dev,
+									gpio_port_value_t *value)
+{
+	struct gpio_davinci_config *config = (struct gpio_davinci_config *)dev->config;
+	volatile struct gpio_davinci_regs *regs =
+				(struct gpio_davinci_regs *)config->port_base;
+
+	*value = regs->in_data;
+
+	return 0;
+}
+
+static int gpio_davinci_port_set_masked_raw(const struct device *dev,
+		gpio_port_pins_t mask, gpio_port_value_t value)
+{
+	struct gpio_davinci_config *config = (struct gpio_davinci_config *)dev->config;
+	volatile struct gpio_davinci_regs *regs =
+				(struct gpio_davinci_regs *)config->port_base;
+
+	regs->out_data = (regs->out_data & (~mask)) | (mask & value);
+
+	return 0;
+}
+
+static int gpio_davinci_port_set_bits_raw(const struct device *dev,
+									gpio_port_pins_t mask)
+{
+	struct gpio_davinci_config *config = (struct gpio_davinci_config *)dev->config;
+	volatile struct gpio_davinci_regs *regs =
+				(struct gpio_davinci_regs *)config->port_base;
+
+	regs->set_data |= mask;
+
+	return 0;
+}
+
+static int gpio_davinci_port_clear_bits_raw(const struct device *dev,
+									gpio_port_pins_t mask)
+{
+	struct gpio_davinci_config *config = (struct gpio_davinci_config *)dev->config;
+	volatile struct gpio_davinci_regs *regs =
+				(struct gpio_davinci_regs *)config->port_base;
+
+	regs->clr_data |= mask;
+
+	return 0;
+}
+
+static int gpio_davinci_port_toggle_bits(const struct device *dev,
+									gpio_port_pins_t mask)
+{
+	struct gpio_davinci_config *config = (struct gpio_davinci_config *)dev->config;
+	volatile struct gpio_davinci_regs *regs =
+				(struct gpio_davinci_regs *)config->port_base;
+
+	regs->out_data ^= mask;
+
+	return 0;
+}
+
+static const struct gpio_driver_api gpio_davinci_driver_api = {
+	.pin_configure = gpio_davinci_configure,
+	.port_get_raw = gpio_davinci_port_get_raw,
+	.port_set_masked_raw = gpio_davinci_port_set_masked_raw,
+	.port_set_bits_raw = gpio_davinci_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_davinci_port_clear_bits_raw,
+	.port_toggle_bits = gpio_davinci_port_toggle_bits
+};
+
+static int gpio_davinci_init(const struct device *dev)
+{
+	return 0;
+}
+
+#define GPIO_J721E_INIT_FUNC(n)	\
```

Comment:
> GPIO_J721E_INIT_FUNC

Are we allowing device specific stuff in a driver? If it's not J721E specific change the naming.